File: lwjgl/src/main/java/com/darkraha/opengldemoj/gl/shader/Texture2DBuilder.java
package com.darkraha.opengldemoj.gl.shader;

import static com.darkraha.opengldemoj.gl.shader.ShaderProgramBuilder.SHADER_TYPE_FRAGMENT;
import static com.darkraha.opengldemoj.gl.shader.ShaderProgramBuilder.SHADER_TYPE_VERTEX;

public class Texture2DBuilder implements HelperBuilder{


    @Override
    public void addTypeDeclarations(StringBuilder sb, int shaderType) {

    }

    @Override
    public void addInputDeclarations(StringBuilder sb, int shaderType) {
        if(shaderType==SHADER_TYPE_VERTEX){
            sb.append("layout(location=3) in vec2 vTexPos ;\n");
        }
    }

    @Override
    public void addExchangeDeclarations(StringBuilder sb, int shaderType) {

        if(shaderType==SHADER_TYPE_VERTEX){
            sb.append( "out vec2 exTexPos;\n");
        }

        if(shaderType==SHADER_TYPE_FRAGMENT){
            sb.append( "in vec2 exTexPos;\n");
        }
    }

    @Override
    public void addUniformDeclarations(StringBuilder sb, int shaderType) {
        if(shaderType==SHADER_TYPE_FRAGMENT){
            sb.append("uniform sampler2D sampler;\n");
        }

    }

    @Override
    public void addFuncsDeclarations(StringBuilder sb, int shaderType) {

    }

    @Override
    public void addCalculations(StringBuilder sb, int shaderType) {

        if(shaderType==SHADER_TYPE_VERTEX){
            sb.append("exTexPos = vTexPos;\n");
        }

        if(shaderType==SHADER_TYPE_FRAGMENT){
            sb.append("fragColor = fragColor*texture(sampler, exTexPos);\n");
        }

    }
}

File: lwjglkt/src/main/kotlin/com/darkraha/opengldemokt/gl/shader/Texture2Dbuilder.kt
package com.darkraha.opengldemokt.gl.shader

import com.darkraha.opengldemokt.gl.shader.ShaderProgramBuilder.Companion.SHADER_TYPE_FRAGMENT
import com.darkraha.opengldemokt.gl.shader.ShaderProgramBuilder.Companion.SHADER_TYPE_VERTEX
import java.lang.StringBuilder

class Texture2DBuilder : HelperBuilder {

    override fun addTypeDeclarations(sb: StringBuilder, shaderType: Int) {}

    override fun addInputDeclarations(sb: StringBuilder, shaderType: Int) {
        if (shaderType == SHADER_TYPE_VERTEX) {
            sb.append("layout(location=3) in vec2 vTexPos ;\n")
        }
    }

    override fun addExchangeDeclarations(sb: StringBuilder, shaderType: Int) {
        if (shaderType == SHADER_TYPE_VERTEX) {
            sb.append("out vec2 exTexPos;\n")
        }
        if (shaderType == SHADER_TYPE_FRAGMENT) {
            sb.append("in vec2 exTexPos;\n")
        }
    }

    override fun addUniformDeclarations(sb: StringBuilder, shaderType: Int) {
        if (shaderType == SHADER_TYPE_FRAGMENT) {
            sb.append("uniform sampler2D sampler;\n")
        }
    }

    override fun addFuncsDeclarations(sb: StringBuilder, shaderType: Int) {}

    override fun addCalculations(sb: StringBuilder, shaderType: Int) {
        if (shaderType == SHADER_TYPE_VERTEX) {
            sb.append("exTexPos = vTexPos;\n")
        }
        if (shaderType == SHADER_TYPE_FRAGMENT) {
            sb.append("fragColor = fragColor*texture(sampler, exTexPos);\n")
        }
    }
}